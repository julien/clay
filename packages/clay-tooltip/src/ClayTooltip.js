import Component from 'metal-component';
import Soy from 'metal-soy';
import {Align} from 'metal-position';
import {Config} from 'metal-state';
import {EventHandler} from 'metal-events';
import {core, object} from 'metal';
import {dom} from 'metal-dom';

import templates from './ClayTooltip.soy.js';

const singleton = Symbol();
const singletonEnforcer = Symbol();

/**
 * Implementation of ClayTooltip.
 * @extends Component
 */
class ClayTooltip extends Component {
	/**
	 * @inheritDoc
	 */
	constructor(enforcer, config, parentElement) {
		if (enforcer !== singletonEnforcer) {
			throw new Error('Use ClayTooltip.init to initialize ClayTooltip');
		}
		super(config, parentElement);
	}

	/**
	 * Returns the instance for this ClayTooltip.
	 * @memberof ClayTooltip
	 * @param {Object=} config An object with the initial values for this
	 *                         component's state.
	 * @param {boolean|string|Element=} parentElement The element where the
	 * component should be rendered. Can be given as a selector or an element.
	 * If `false` is passed, the component won't be rendered automatically
	 * after created.
	 * @static
	 * @return {ClayTooltip}
	 */
	static init(config, parentElement) {
		if (!this[singleton]) {
			this[singleton] = new ClayTooltip(
				singletonEnforcer,
				config,
				parentElement
			);
		}
		return this[singleton];
	}

	/**
	 * @inheritDoc
	 */
	created() {
		this._eventHandler = new EventHandler();

		for (let selector of this.selectors) {
			this._eventHandler.add(
				dom.delegate(
					document,
					'focus',
					selector,
					this._handleMouseEnter.bind(this)
				),
				dom.delegate(
					document,
					'mouseenter',
					selector,
					this._handleMouseEnter.bind(this)
				),
				dom.delegate(
					document,
					'mouseleave',
					selector,
					this._handleMouseLeave.bind(this)
				),
				dom.delegate(
					document,
					'blur',
					selector,
					this._handleMouseLeave.bind(this)
				)
			);
		}
	}

	/**
	 * @inheritDoc
	 */
	detached() {
		this._eventHandler.removeAllListeners();
	}

	/**
	 * @inheritDoc
	 */
	disposed() {
		this._eventHandler.removeAllListeners();
	}

	/**
	 * Handles mouseenter events.
	 * @memberof ClayTooltip
	 * @param {Object} event The event object.
	 * @protected
	 */
	_handleMouseEnter(event) {
		const target = event.target;
		const content =
			target.getAttribute('data-title') || target.getAttribute('title');

		this._content = content;
		this.alignedPosition = Align.align(this.element, target, this.position);
		this._showTooltip = true;
	}

	/**
	 * Handles mouseleave events.
	 * @memberof ClayTooltip
	 * @param {Object} event The event object.
	 * @protected
	 */
	_handleMouseLeave() {
		this._showTooltip = false;
	}

	/**
	 * The setter function for the `classMap` staet.
	 * @param {Object} val
	 * @return {!Object}
	 * @protected
	 */
	setterClassMapFn_(val) {
		return object.mixin(this.valueClassMapFn_(), val);
	}

	/**
	 * Gets the default value for the `classMap` state.
	 * @return {!Object}
	 * @protected
	 */
	valueClassMapFn_() {
		return {
			[Align.TopLeft]: 'clay-tooltip-top-left',
			[Align.TopCenter]: 'clay-tooltip-top',
			[Align.TopRight]: 'clay-tooltip-top-right',
			[Align.BottomLeft]: 'clay-tooltip-bottom-left',
			[Align.BottomCenter]: 'clay-tooltip-bottom',
			[Align.BottomRight]: 'clay-tooltip-bottom-right',
			[Align.RightCenter]: 'clay-tooltip-right',
			[Align.LeftCenter]: 'clay-tooltip-left',
		};
	}
}

/**
 * State definition.
 * @static
 * @type {!Object}
 */
ClayTooltip.STATE = {
	/**
	 * Content of the tooltip.
	 * @default undefined
	 * @instance
	 * @memberof ClayTooltip
	 * @type {!(html|string)}
	 */
	_content: Config.any()
		.value('')
		.internal(),

	_showTooltip: Config.bool().value(false).internal(),

	/**
	 * The current position of the tooltip after being aligned via `Align.align`.
	 * @default undefined
	 * @type {number}
	 * @instance
	 * @memberof ClayTooltip
	 */
	alignedPosition: {
		validator: Align.isValidPosition,
	},

	/**
	 * Element classes.
	 * @default ''
	 * @type {string}
	 * @instance
	 */
	elementClasses: Config.string().value(''),

	/**
	 * A map from `Align` position constants to the CSS class that should be
	 * added to the tooltip when it's aligned in that position.
	 * @type {!Object}
	 */
	classMap: {
		setter: 'setterClassMapFn_',
		validator: core.isObject,
		valueFn: 'valueClassMapFn_',
	},

	/**
	 * Tooltip position
	 * @default Align.Bottom
	 * @instance
	 * @memberof ClayTooltip
	 * @type {!string}
	 */
	position: {
		validator: Align.isValidPosition,
		value: Align.Bottom,
	},

	/**
	 * Selectors
	 * @default ['[data-title]', '[title]']
	 * @instance
	 * @memberof ClayTooltip
	 * @type {!Array.<string>}
	 */
	selectors: Config.array().value(['[data-title]', '[title]']),

	/**
	 * Trigger events
	 * @default ['[mouseenter]', '[mouseleave]']
	 * @instance
	 * @memberof ClayTooltip
	 * @type {!Array.<string>}
	 */
	triggerEvents: Config.array().value(['focus', 'mouseenter', 'mouseleave']),
};

Soy.register(ClayTooltip, templates);

export {ClayTooltip};
export default ClayTooltip;
