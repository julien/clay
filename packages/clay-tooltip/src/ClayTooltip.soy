{namespace ClayTooltip}

/**
 * This renders a tooltip.
 */
{template .render}
	{@param? _content: html|string}
	{@param? _showTooltip: bool}
	{@param? alignedPosition: number}
	{@param? classMap: ?}
	{@param? elementClasses: string}
	{@param? position: number}

	{let $classes: $classMap ? $classMap : ['clay-tooltip-top-left', 'clay-tooltip-top', 'clay-tooltip-top-right', 'clay-tooltip-bottom-left', 'clay-tooltip-bottom', 'clay-tooltip-bottom-right', 'clay-tooltip-right', 'clay-tooltip-left'] /}
	{let $currentPosition: $alignedPosition ?: $position /}
	{let $positionClass: isNonnull($currentPosition) ? $classes[$currentPosition] : 'clay-tooltip-bottom' /}

	{let $tooltipAttributes kind="attributes"}
		class="tooltip
			{sp}{$positionClass}

			{if $elementClasses}
				{sp}{$elementClasses}
			{/if}

			{if $_showTooltip}
				{sp}show
			{/if}			
		"
		role="tooltip"
	{/let}

	<div {$tooltipAttributes}>
		<div class="arrow"></div>
	
		<div class="tooltip-inner">
			{$_content}
		</div>
	</div>
{/template}
