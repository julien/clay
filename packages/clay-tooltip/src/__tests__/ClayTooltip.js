import ClayTooltip from '../ClayTooltip';

let clayTooltip;

describe('ClayTooltip', function() {
	afterEach(() => {
		if (clayTooltip) {
			clayTooltip.dispose();
		}
	});

	it('should render a tooltip with content', () => {
		clayTooltip = ClayTooltip.init();
		expect(clayTooltip).toMatchSnapshot();
	});
});
