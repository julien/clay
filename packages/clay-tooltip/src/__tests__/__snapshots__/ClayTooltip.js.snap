// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClayTooltip should render a tooltip with content 1`] = `
<div class="tooltip clay-tooltip-bottom " role="tooltip">
  <div class="arrow"></div>
  <div class="tooltip-inner"></div>
</div>
`;
